govcd/saml_auth.go

```diff
@@ -0,0 +1,292 @@
+	}
+
+	return nil
+}
+
+// vcdAuthorizeSamlGetAdfsServer finds out Active Directory Federation Service (ADFS) server to use
+// for SAML authentication
+// It works by temporarily patching existing http.Client behavior to avoid automatically
+// following HTTP redirects and searches for Location header after the request to vCD SAML redirect
+// address. The URL to search redirect location is:
+// url.Scheme + "://" + url.Host + "/login/my-org/saml/login/alias/vcd?service=tenant:" + org
+func vcdAuthorizeSamlGetAdfsServer(vcdCli *VCDClient, org string) (string, error) {
+	url := vcdCli.Client.VCDHREF
+
+	// Backup existing http.Client redirect behavior so that it does not follow HTTP redirects
+	// automatically and restore it right after this function by using defer. A new http.Client
+	// could be spawned here, but the existing one is re-used on purpose to inherit all other
+	// settings used for client (timeouts, etc).
+	backupRedirectChecker := vcdCli.Client.Http.CheckRedirect
+
+	defer func() {
+		vcdCli.Client.Http.CheckRedirect = backupRedirectChecker
+	}()
+
+	// Patch http client to avoid following redirects
+	vcdCli.Client.Http.CheckRedirect = func(req *http.Request, via []*http.Request) error {
+		return http.ErrUseLastResponse
+	}
+
+	// Construct SAML login URL which should return a redirect to ADFS server
+	loginURLString := url.Scheme + "://" + url.Host + "/login/" + org + "/saml/login/alias/vcd"
+	loginURL, err := url.Parse(loginURLString)
+	if err != nil {
+		return "", fmt.Errorf("unable to parse login URL '%s': %s", loginURLString, err)
+	}
+	util.Logger.Printf("[DEBUG] SAML looking up IdP (ADFS) host redirect in: %s", loginURL.String())
+
+	// Make a request to URL adding unencoded query parameters in the format:
+	// "?service=tenant:my-org"
+	req := vcdCli.Client.NewRequestWitNotEncodedParams(
+		nil, map[string]string{"service": "tenant:" + org}, http.MethodGet, *loginURL, nil)
+	httpResponse, err := checkResp(vcdCli.Client.Http.Do(req))
+	if err != nil {
+		return "", fmt.Errorf("SAML - ADFS server query failed: %s", err)
+	}
+
+	err = decodeBody(httpResponse, nil)
+	if err != nil {
+		return "", fmt.Errorf("SAML - error decoding body: %s", err)
+	}
+
+	// httpResponse.Location() returns an error if no 'Location' header is present
+	adfsEndpoint, err := httpResponse.Location()
+	if err != nil {
+		return "", fmt.Errorf("SAML GET request for '%s' did not return HTTP redirect. "+
+			"Is SAML configured? Got error: %s", loginURL, err)
+	}
+
+	authEndPoint := adfsEndpoint.Scheme + "://" + adfsEndpoint.Host + "/adfs/services/trust/13/usernamemixed"
+	util.Logger.Printf("[DEBUG] SAML got IdP login endpoint: %s", authEndPoint)
+
+	return authEndPoint, nil
+}
+
+// vcdAuthorizeSamlGetSamlEntityId attempts to load vCD hosted SAML metadata from URL:
+// url.Scheme + "://" + url.Host + "/cloud/org/" + org + "/saml/metadata/alias/vcd"
+// Returns an error if Entity ID is empty
```

Comment:
I could, but these bodies are not tiny. This metadata body will be like 100lines.

Reply:
I guess, but somehow we should keep knowledge in house, otherwise if I have to investigate or make simple change and won't have Adf server, then will be impossible.

Reply:
You can still find them in mock server sample responses: https://github.com/vmware/go-vcloud-director/pull/304/files#diff-8654349600100b8c06dd2eeae8018962R1193

Reply:
So maybe you just could say where to found response example in comment?

Reply:
Added comment